Create backups on stand-by masters

ggezer opened the pull request:
This is to add a capability for delegating backups to stand-by masters in HA cluster. Hence, avoiding the need to pause metadata changes on leader during the whole backup.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/6244/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/6245/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/6246/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/6275/
Test FAILed.

LuQQiu commented on core/server/master/src/main/java/alluxio/master/backup/BackupLeaderRole.java:
```diff
@@ -0,0 +1,293 @@
+        adjustAbandonTimeout(false);
+        return true;
+      } catch (Exception e) {
+        LOG.warn(String.format("Failed to delegate backup to worker: %s", connection), e);
```
The same. in Alluxio logs, user may think it's Alluxio worker instead of backup worker.

LuQQiu commented on core/server/master/src/main/java/alluxio/master/backup/BackupWorkerRole.java:
```diff
@@ -0,0 +1,369 @@
+    Preconditions.checkState(!mBackupTracker.inProgress(), "Backup in progress");
+
+    // Cancel timeout timer.
+    mSuspendTimeoutFuture.cancel(true);
```
Is it possible that this worker failed to receive the suspend message but receive the request message?
Should we add a preconditions.checkState or add a null check?

ggezer replied:
Backup-leader don't send request before sending suspend message.

LuQQiu commented on core/server/common/src/main/java/alluxio/master/journal/raft/JournalStateMachine.java:
```diff
@@ -303,4 +322,157 @@ public boolean isSnapshotting() {
+        return CompletableFuture.completedFuture(null);
+      }
+      // Create a future for completion.
+      mAdvanceFuture = new CompletableFuture<>();
```
Looks like the advance behavior is synchronous, 
it will always be   
```
mAdvanceFuture = new CompletableFuture<>();
while loop to applyToMaster() -> mAdvanceFuture.complete(null);
return advanceFuture
```
Why do we need to use a completableFuture?

ggezer replied:
It's synchronous only for the part that is in the buffer. If target sequence is not met, then future will not be completed until state machine receives it.

LuQQiu replied:
Got it! Thanks

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/6320/
Test FAILed.

ggezer commented on core/server/master/src/main/java/alluxio/master/backup/BackupLeaderRole.java:
```diff
@@ -0,0 +1,293 @@
+    services
+        .put(ServiceType.META_MASTER_BACKUP_MESSAGING_SERVICE,
+            new GrpcService(
+                ServerInterceptors.intercept(
```
`ClientIpAddressInjector` is being injected to every call into`CopycatMessageServiceClientHandler` here. 

It's used mostly for traces. Otherwise It's very hard to say who's connecting to server in a gRPC bi-di stream.

LuQQiu replied:
Got it, thanks

ggezer commented on core/server/master/src/main/java/alluxio/master/backup/BackupLeaderRole.java:
```diff
@@ -0,0 +1,293 @@
+        .put(ServiceType.META_MASTER_BACKUP_MESSAGING_SERVICE,
+            new GrpcService(
+                ServerInterceptors.intercept(
+                    new CopycatMessageServiceClientHandler(
```
I'll follow up with another PR for refactoring these classes for more generic use. Once this particular usage is locked then I can have better idea for refactoring.

ggezer commented on core/common/src/main/java/alluxio/conf/PropertyKey.java:
```diff
@@ -1194,19 +1194,47 @@ public String toString() {
-          .setDescription("The hostname that Alluxio master binds to.")
+  public static final PropertyKey MASTER_BACKUP_DELEGATION_ENABLED =
+      new Builder(Name.MASTER_BACKUP_DELEGATION_ENABLED)
+          .setDefaultValue(false)
```
In non-HA cluster, we'll allow backups on leader due to below check.
` boolean delegateBackup =
      ServerConfiguration.getBoolean(PropertyKey.MASTER_BACKUP_DELEGATION_ENABLED)
            **&& ConfigurationUtils.isHaMode**(ServerConfiguration.global());`

Only in HA cluster, we won't allow blocking the leader by default.

That being said, we might still want to allow old way of backups in HA cluster. That's open to discussion.

LuQQiu replied:
I try to understand how we backup in this PR:
(1) single master -> backup on local (allow_leader option is not needed)
(2) HA mode  -> delegation enabled -> backup on followers
                        -> delegation enabled -> but don't have follower connections -> check if backup allowed on leader -> Yes:backup on leader  No:error out
                        -> delegation not enabled   -> backup on leader

So the option allow_leader only affects HA cluster with delegation enabled but don't have any follower connections. 
I previously think allow-leader==false means we don't allow leader to take backup even in single master mode.  

```
  private static final Option ALLOW_LEADER_OPTION =
      Option.builder()
          .longOpt("allow-leader")
          .required(false)
          .hasArg(false)
          .desc("whether to allow leader to take the backup when there are no secondaries.")
          .build();
```
Can we change the description to let users know this option only affects high availability mode. In single master mode, backup is always taken in the master

ggezer commented on shell/src/main/java/alluxio/cli/fsadmin/command/BackupCommand.java:
```diff
@@ -41,6 +42,13 @@
+          .longOpt("allow-leader")
+          .required(false)
+          .hasArg(false)
+          .desc("whether to allow leader to take the backup when there are no secondaries.")
```
@apc999?

ggezer replied:
I think this should be stand-by master for consistency with other documents.

ggezer commented on shell/src/main/java/alluxio/cli/fsadmin/command/BackupCommand.java:
```diff
@@ -56,33 +64,64 @@ public String getCommandName() {
+      }
+      if (status.getState() == BackupState.COMPLETED) {
+        break;
+      } else if (status.getState() == BackupState.FAULTED) {
```
Proto doesn't allow it since`meta_master.proto` has another enum that defined `FAILED`. 
I can prefix all values with `Backup_`. Or just leave it as this if not that ugly.

LuQQiu replied:
So strange, but simple is better, let's maintain FAULTED

ggezer commented on core/common/src/main/java/alluxio/conf/PropertyKey.java:
```diff
@@ -1194,19 +1194,47 @@ public String toString() {
+          .setConsistencyCheckLevel(ConsistencyCheckLevel.IGNORE)
+          .setScope(Scope.MASTER)
+          .build();
+  public static final PropertyKey MASTER_BACKUP_ABANDON_TIMEOUT_MS =
```
CONNECTION_TIMEOUT might not be a perfect fit as it's misleading in other ways. This property basically controls timing out of backup after not getting heart-beats for it.

LuQQiu replied:
Can you remove all the `_MS` from the property key? also for their names

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/6340/
Test FAILed.

ggezer commented:
@calvinjia @LuQQiu 
For reviewers' reference, this work will be followed by 3 PRs:

**1. Refactoring copycat transport:** Now that it's used by two separate features, it should be hosted by an appropriate component and refactored to generalize copycat specific names/traces etc...

**2. JournalStateMachine concurrency:** `JournalStateMachine` is a globally locked class and this PR doesn't aim to change that. It also added an additional contention when resuming journals. If locking during `resume()` takes a long time, then quorum could potentially be disrupted. I was not able to drive system to that point in manual tests but it's very much possible under heavy load.

**3. On-disk buffering while RAFT is suspended:** Currently entries are buffered on memory while journal is suspended. Buffer should be extended to disk.

LuQQiu commented on core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournal.java:
```diff
@@ -217,11 +225,24 @@ public synchronized void start() throws IOException {
+    if (mSuspended) {
+      // Cancel if advancing is in progress.
+      if (mAdvanceFuture != null && !mAdvanceFuture.isDone()) {
+        // TODO(ggezer): Avoid re-applying sequences read while advancing.
```
Did we stop backing up?

ggezer replied:
Yes, backup will be cancelled if backup-worker fail over to become a leader.

LuQQiu commented on core/server/master/src/main/java/alluxio/master/backup/BackupWorkerRole.java:
```diff
@@ -0,0 +1,369 @@
+  private void enforceResumeJournals() {
+    try {
+      mJournalSystem.resume();
+    } catch (IOException e) {
```
catch throwable instead?

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/6495/
Test PASSed.

LuQQiu commented on core/server/master/src/main/java/alluxio/master/backup/BackupWorkerRole.java:
```diff
@@ -215,6 +220,11 @@ private void handleRequestMessage(BackupRequestMessage requestMsg) {
           LOG.info("Journal transition completed. Scheduling a backup.");
+          if (!mBackupTimeoutTask.cancel(true)) {
+            mBackupTracker
+                .updateError(new BackupException("Journal has ben resumed due to time-out."));
```
When the advance future being cancelled, journal will be automatically resume?

ggezer replied:
It should.
However it looks like CompletableFuture<?> does not support cancelling. So I'll need to change it in the signature to Future<?> and re-implement related parts.

ggezer replied:
@LuQQiu  `CompletableFuture<?>` was not cancelleable. 

`JournalSystem#advance` now returns `AdvanceFuture` which allows for graceful termination of ongoing advancing.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/6497/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/6539/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/6544/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/6548/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/6561/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/6581/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/6584/
Test PASSed.

ggezer commented:
@calvinjia @LuQQiu PTAL. Thanks!

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/6586/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/6590/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/6583/
Test FAILed.

ggezer commented:
Jenkins, test this please

LuQQiu commented:
@ggezer What's the major thing that you are fixing for delegation backup recently?  Because of the force push, I am not quite sure what's the major changes

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/6592/
Test PASSed.

LuQQiu commented on core/server/common/src/main/java/alluxio/master/journal/AbstractAdvanceThread.java:
```diff
@@ -0,0 +1,55 @@
+    try {
+      runAdvance();
+    } catch (Exception e) {
+      ProcessUtils.fatalError(LOG, e, "Advance thread is failed.");
```
I try to list the workflow:
```
advancing
When BACKUP_ABORT_AFTER_TRANSITION_TIMEOUT_MS exceed
CompletableFuture will be cancel/interrupted
runAdvance will throw runtime exception
here will make it as a fatal error
system exit
```
Is it what's happening?

ggezer replied:
`CompletableFuture` in `BackupWorkerRole#handleRequestMessage` is used to track `AdvanceFuture#waitTermimation`. Cancelling it won't interrupt advancing. 

Also `AdvanceFuture#cancel` itself won't interrupt thread but just mark for termination.

So timing out on advance won't trigger interrupt and will simply call `resume`.Which will cause advancing to finish prematurely.

I can implement `AdvanceFuture#waitTermination(long timeout)` variation to unclutter this piece of code but that itself will be complicated since `AdvanceFuture` can hold multiple `AdvanceFuture`s

ggezer replied:
@LuQQiu Do you have any more concerns around this?

calvinjia commented on core/server/common/src/main/java/alluxio/master/journal/raft/BufferedJournalApplier.java:
```diff
@@ -0,0 +1,293 @@
+  private long mLastAppliedSequence = -1;
+
+  /** Whether this state machine is suspended. */
+  private boolean mSuspended = false;
```
Could we do with having an atomic boolean instead of using the locks?

ggezer replied:
`Resume` makes it complicated since it needs to actually lock.

calvinjia commented on core/server/common/src/main/java/alluxio/master/journal/raft/BufferedJournalApplier.java:
```diff
@@ -0,0 +1,293 @@
+      Preconditions.checkState(mAdvanceTask == null || !mAdvanceTask.isAlive(),
+          "Advance task in progress.");
+      Preconditions.checkState(sequence >= 0, "Invalid sequence");
+      Preconditions.checkState(mLastAppliedSequence <= sequence, "Can't advance to past.");
```
Print out the sequence numbers here?

calvinjia commented on core/server/common/src/main/java/alluxio/master/journal/raft/BufferedJournalApplier.java:
```diff
@@ -0,0 +1,293 @@
+      Preconditions.checkState(!mResumeInProgress, "Resume in progress");
+      Preconditions.checkState(mAdvanceTask == null || !mAdvanceTask.isAlive(),
+          "Advance task in progress.");
+      Preconditions.checkState(sequence >= 0, "Invalid sequence");
```
Print the invalid sequence number?

calvinjia commented on core/server/common/src/main/java/alluxio/master/journal/raft/BufferedJournalApplier.java:
```diff
@@ -0,0 +1,293 @@
+  private boolean mResumeInProgress = false;
+
+  /** Buffer to use during suspend period. */
+  private Queue<Journal.JournalEntry> mSuspendBuffer = new ConcurrentLinkedQueue();
```
Do we need to use the Concurrent object if we are managing the concurrency with locks? The size operation is more expensive.

ggezer replied:
It's possible for `resume` and `processJournalEntry` to be accessing concurrently. This is until `resume` decides to take the lock in order to complete a long running catching up session.

ggezer replied:
`AdvanceThread` also synchronizes with monitors only for getting notified when there are entries in the buffer, it doesn't lock while popping entries.

calvinjia commented on core/server/common/src/main/java/alluxio/master/journal/raft/RaftJournalSystem.java:
```diff
@@ -365,6 +369,40 @@ public synchronized void losePrimacy() {
+  public synchronized void resume() throws IOException {
+    mStateMachine.resume();
+    // Resume copycat snapshots.
+    mSnapshotAllowed.set(true);
```
Is it safe to allow snapshots?

ggezer replied:
The whole suspend/resume cycle is reserved for secondary role. So this is just restoring the original value after the backup.

calvinjia commented on core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournal.java:
```diff
@@ -104,6 +106,15 @@
    */
   private UfsJournalCheckpointThread mTailerThread;
 
+  /** Whether the journal was suspended. */
```
`was` -> `is`?

calvinjia commented on core/server/master/src/main/java/alluxio/master/backup/AbstractBackupRole.java:
```diff
@@ -0,0 +1,209 @@
+    AlluxioURI backupUri;
+    // Acquire the UFS resource under which backup is being created.
+    try (CloseableResource<UnderFileSystem> ufsResource =
+        mUfsManager.getRoot().acquireUfsResource()) {
```
Are backups only written to the root ufs?

ggezer replied:
Nice catch, I think a fix for that regression was checked in after I started this.

ggezer replied:
Oh, Looks like I still have that fix. In line 152 it overrides `ufs` resource with local factory if local backup is requested.

calvinjia replied:
Do we allow the case where it is a nonlocal, nonroot UFS?

ggezer replied:
https://github.com/Alluxio/alluxio/issues/10066

calvinjia commented on core/server/master/src/main/java/alluxio/master/backup/BackupWorkerRole.java:
```diff
@@ -0,0 +1,369 @@
+      LOG.info("Suspended journals for backup.");
+    } catch (IOException e) {
+      String failMessage = "Failed to suspended journals for backup.";
+      LOG.error(failMessage, e);
```
Do we need to log an error? We are already throwing an exception.

ggezer replied:
Exceptions during message handling will show up only on the receiver side. This is to ease correlating.

calvinjia commented on core/server/master/src/main/java/alluxio/master/backup/BackupWorkerRole.java:
```diff
@@ -0,0 +1,369 @@
+        if (mBackupTracker.isFinished()) {
+          break;
+        }
+      } while (true);
```
Any reason not to use `do...while` instead of just `while`?

ggezer commented on core/server/common/src/main/java/alluxio/master/journal/JournalSystem.java:
```diff
@@ -138,6 +139,40 @@
+   *
+   * @return the current map of sequences for each master
+   */
+  Map<String, Long> getCurrentSequences();
```
Yes. Do you have a suggestion for a better name?

calvinjia replied:
`getCurrentSequenceNumbers`?

ggezer commented on core/server/common/src/main/java/alluxio/master/journal/AdvanceFuture.java:
```diff
@@ -0,0 +1,82 @@
+   * @param futures list of advance futures
+   * @return unified advance future
+   */
+  public static AdvanceFuture allOf(List<AdvanceFuture> futures) {
```
I mimicked `CompletableFuture` in that. Static creator is to combine multiple instances together.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/6668/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/6690/
Test PASSed.

calvinjia commented on core/common/src/main/java/alluxio/wire/BackupStatus.java:
```diff
@@ -0,0 +1,225 @@
+/**
+ * Defines the status of a backup.
+ */
+public class BackupStatus {
```
Should this class be immutable?

ggezer replied:
It's used as a wire type. I think it's OK to leave it mutable as with other types. Unless you have a particular reason?

calvinjia replied:
Oh I see, it would be nice to make our wire types immutable, but I'm fine with this since its the way we do it for our other ones.

calvinjia commented on core/common/src/main/java/alluxio/wire/BackupStatus.java:
```diff
@@ -0,0 +1,225 @@
+   * @return {@code true} if the backup is failed
+   */
+  public boolean isFailed() {
+    return mState == BackupState.FAULTED;
```
Should the stated be `FAILED`?

ggezer replied:
I could only make `Failed` as there is another enum in the same proto with value `FAILED`.

calvinjia replied:
Oh I see, most projects get around this by naming the enums like `BackupState.BACKUP_STATE_FAILED`

ggezer replied:
I just didn't want to lengthen enum values. We can do that on the next 'Failed' value in the same file.

calvinjia commented on core/server/common/src/main/java/alluxio/master/journal/AdvanceFuture.java:
```diff
@@ -0,0 +1,82 @@
+/**
+ * Used to track and cancel journal advancing.
+ */
+public class AdvanceFuture {
```
Naming suggestion, does `catchup` make more sense here?

ggezer replied:
For this class only or does you suggestion includes `JournalSystem.advance()` as well?

calvinjia replied:
Replace the concept of advance with catchup since we already have a concept of catchup.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/6696/
Test PASSed.

calvinjia commented on core/common/src/main/java/alluxio/wire/BackupStatus.java:
```diff
@@ -0,0 +1,225 @@
+ * Defines the status of a backup.
+ */
+public class BackupStatus {
+  /** State of backup. */
```
Should there be an unique identifier for the backup?

ggezer replied:
I was planning on getting unique IDs after uncovering some client side issue.
When run concurrently, backup shell commands could start picking up results for runs that they haven't launched.

ggezer replied:
https://github.com/Alluxio/alluxio/pull/10182/commits/2ed4f85af8905341a0b972c38f62d49e921f3ce9

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/6707/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/6724/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/6729/
Test FAILed.

ggezer commented:
Jenkins, test this please

ggezer commented on shell/src/main/java/alluxio/cli/fsadmin/command/BackupCommand.java:
```diff
@@ -97,12 +136,13 @@ public String getDescription() {
         + " master, use --local and specify a filesystem path. Backing up metadata"
-        + " requires a pause in master metadata changes, so use this command sparingly to"
-        + " avoid interfering with other users of the system.";
+        + " will be delegated to secondary masters in HA cluster. Use --allow-leader for"
```
I think this should be stand-by master for consistency with other documents.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/6730/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/6731/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/6732/
Test FAILed.

ggezer commented:
Jenkins, test this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/6734/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/6735/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/6746/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/6747/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/6783/
Test PASSed.

LuQQiu commented on core/server/common/src/main/java/alluxio/master/journal/raft/BufferedJournalApplier.java:
```diff
@@ -0,0 +1,295 @@
+        synchronized (mSuspendBuffer) {
+          while (!mStopCatchingUp && mSuspendBuffer.size() == 0) {
+            try {
+              mSuspendBuffer.wait();
```
I am not sure how `ConcurrentLinkedQueue` work with `Synchronized`.
When running catchup, we 
```
        synchronized (mSuspendBuffer) {
          while (!mStopCatchingUp && mSuspendBuffer.size() == 0) {
            try {
              mSuspendBuffer.wait();
            } catch (InterruptedException e) {
              throw new RuntimeException("Interrupted while catching up.");
            }
          }
```
In this situation, will other `mSuspendBuffer.notifyAll()` operations be blocked by not able to get the `synchronized (mSuspendBuffer)`?

ggezer replied:
Buffer could be accessed by `processJournalEntry`, `resume`, `catchup`.
`processJournalEntry` can be called concurrently with `resume` OR `catchup`. All three won't run in parallel since `resume` first waits until active `catchup` is done.

For when `processJournalEntry` and `catchup` is running concurrently. A monitor is required to get notified when the queue has an entry, so that `catchup` can continue. Note that resume locks only to check if there are entries, so the monitor won't get that hot. For the monitor I used the buffer itself, which will have no effect on concurrency of queue puts/gets.

For when, `processJournalEntry` and `resume` is running concurrently. We don't want to block anything until `resume` decides that it's time to block `processJournalEntry` and finish resuming.

ggezer commented on core/server/common/src/main/java/alluxio/master/journal/raft/BufferedJournalApplier.java:
```diff
@@ -0,0 +1,295 @@
+     *
+     * It doesn't block state until
+     *   -> buffer contains few elements ( RESUME_LOCK_BUFFER_SIZE_WATERMARK )
+     *   -> was running for a long time  ( RESUME_LOCK_TIME_LIMIT_MS         )
```
`processJournalEntry` could always cut in.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/6903/
Test PASSed.

calvinjia commented on core/server/common/src/main/java/alluxio/master/journal/raft/BufferedJournalApplier.java:
```diff
@@ -0,0 +1,298 @@
+import java.util.function.Supplier;
+
+/**
+ * Used to control applying to masters.
```
Could you add some more info to the javadoc here (ie. we are applying journal entries and might need to buffer them if we are suspended)?

calvinjia commented on core/server/common/src/main/java/alluxio/master/journal/raft/BufferedJournalApplier.java:
```diff
@@ -0,0 +1,298 @@
+  private AbstractCatchupThread mCatchupThread;
+
+  /** Used to synchronize buffer state. */
+  private ReentrantLock mStateLock = new ReentrantLock(true);
```
Any reason for making this a fair lock?

ggezer replied:
A guard against aggressive `processJournalEntry` calls by copycat.

calvinjia commented on core/server/common/src/main/java/alluxio/master/journal/raft/BufferedJournalApplier.java:
```diff
@@ -0,0 +1,298 @@
+      Preconditions.checkState(!mSuspended, "Already suspended");
+      LOG.info("Suspending state machine.");
+      mSuspended = true;
+      LOG.info("Suspended state machine at sequence: {}", mLastAppliedSequence);
```
The previous log message can be removed since there is trivial logic in between.

calvinjia commented on core/server/common/src/main/java/alluxio/master/journal/raft/BufferedJournalApplier.java:
```diff
@@ -0,0 +1,298 @@
+      Preconditions.checkState(mSuspended, "Not suspended");
+      Preconditions.checkState(!mResumeInProgress, "Resume in progress");
+      mResumeInProgress = true;
+      LOG.info("Resuming state machine.");
```
Log the last applied sequence number?

calvinjia commented on core/server/common/src/main/java/alluxio/master/journal/raft/BufferedJournalApplier.java:
```diff
@@ -0,0 +1,298 @@
+     */
+    try {
+      // Mark resume start time.
+      Instant resumeStartTime = Instant.now();
```
Users have brought up performance issues with `Instant.now`, can we use `System.nanoTime`?

calvinjia commented on core/server/common/src/main/java/alluxio/master/journal/raft/BufferedJournalApplier.java:
```diff
@@ -0,0 +1,298 @@
+        boolean lockSubmission = !mStateLock.isHeldByCurrentThread()
+            && (mSuspendBuffer.size() <= RESUME_LOCK_BUFFER_SIZE_WATERMARK
+                || Duration.between(Instant.now(), resumeStartTime)
+                    .compareTo(Duration.ofMillis(RESUME_LOCK_TIME_LIMIT_MS)) > 0);
```
Is there a downside to always grabbing the lock and flushing the entire queue? In the worst case scenario, this is what we revert to after waiting `RESUME_LOCK_TIME_LIMIT`

ggezer replied:
This is to not prevent `processJournalEntry` if buffer was overloaded during a long backup. Blocking it for a long time would render copycat server unreachable.

calvinjia replied:
Could we still be in that situation after `RESUME_LOCK_TIME_LIMIT` is up?

ggezer replied:
Right. Hitting a time limit is when the code stops being hopeful. Hitting a time limit is extremely unlikely because resuming is done on a thread, where as new entries will come in as part of copycat hearbeats.

calvinjia commented on core/server/common/src/main/java/alluxio/master/journal/raft/BufferedJournalApplier.java:
```diff
@@ -0,0 +1,298 @@
+  private final Supplier<Set<JournalSink>> mJournalSinks;
+
+  /** The last sequence applied. */
+  private long mLastAppliedSequence = -1;
```
Do we need to track one sequence number for each journal?

ggezer replied:
This is used by `JournalStateMachine` which manages all journals with a single sequence. So not required at here as well.

calvinjia commented on core/server/master/src/main/java/alluxio/master/backup/BackupWorkerRole.java:
```diff
@@ -0,0 +1,378 @@
+
+    // Cancel timeout task created by suspend message handler.
+    if (!mBackupTimeoutTask.cancel(true)) {
+      mBackupTracker.updateError(new BackupException("Journal has ben resumed due to a time-out"));
```
`ben`->`been`

calvinjia commented on core/server/master/src/main/java/alluxio/master/backup/BackupWorkerRole.java:
```diff
@@ -0,0 +1,378 @@
+   */
+  private void enforceResumeJournals() {
+    try {
+      mJournalSystem.resume();
```
Is there a timing issue where some journals could have been resumed and we cancel the task, or will that cause an exception to be thrown here?

ggezer replied:
There is no such window AFAICT.

Double resuming should be avoided by this line below and the fact that, gaining-primacy first stops this role...

`// Cancel timeout task created by suspend message handler.
    if (!mBackupTimeoutTask.cancel(true)) {
      mBackupTracker.updateError(new BackupException("Journal has been resumed due to a time-out"));
      return;
    }`

calvinjia replied:
`mBackupTimeoutTask` can return true if it interrupts `mJournalSystem.resume` while it is in progress right?

ggezer replied:
Reverted try-catch condition in `enforceResumeJournal` to original.

calvinjia commented on core/server/master/src/main/java/alluxio/master/backup/BackupWorkerRole.java:
```diff
@@ -0,0 +1,378 @@
+        mBackupTracker.updateError(new BackupException(
+            String.format("Failed to catch-up journals: %s", e.getMessage()), e));
+        // Don't resume journals if interrupted.
+        // Catch-up future can be interrupted only when master becomes primary during backup.
```
Is this true, don't we cancel the future if we lose connection to the master?

ggezer replied:
Nice catch!

This statement is not correct:

First; the future that is being interrupted is not the `catchup-future` but the async task that is spawned to wait for `catchup-future`. (Waiting on catchup-future is not interruptable)
Second; we interrupt this code when connection to master is lost.

The reason, I didn't enforce resuming when interrupted was to quit from stopMasters() path ASAP during gain-primacy scenario. However the journal itself will still call it when gaining primacy, so being quick here does not really have a value. So removing this code completely.

ggezer commented on core/client/fs/src/main/java/alluxio/client/meta/MetaMasterClient.java:
```diff
@@ -12,30 +12,48 @@
+   * @return status of backup
+   * @throws IOException
+   */
+  BackupStatus backup(BackupPRequest backupRequest) throws IOException;
```
Currently, backup will return complete BackupStatus unless you specify -async flag in the request. Are you suggesting that it returns UUID even when it was blocking?

calvinjia replied:
Oh I see, that makes sense if its synchronous. Does this mean if there is an error, an exception will not be thrown but a backup status with an error will be returned?

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/6915/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/6917/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/6922/
Test PASSed.

ggezer commented:
alluxio-bot, merge this please
